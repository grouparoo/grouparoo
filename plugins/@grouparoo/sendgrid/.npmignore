# General
node_modules
.DS_Store
.vscode
npm-debug.log
yarn-error.log
.env
dump.rdb
*.retry
coverage
