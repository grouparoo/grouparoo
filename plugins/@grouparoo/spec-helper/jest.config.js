const { helper } = require("./dist/index");

module.exports = {
  testEnvironment: "node",
  testTimeout: helper.defaultTime,
  maxWorkers: "50%",
  setupFiles: ["<rootDir>/jest.setup.js"],
  transform: {
    "^.+\\.ts?$": "ts-jest",
  },
  testPathIgnorePatterns: ["<rootDir>/src", "<rootDir>/dist"],
};
