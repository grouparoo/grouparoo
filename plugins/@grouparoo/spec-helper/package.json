{
  "author": "Grouparoo Inc <hello@grouparoo.com>",
  "name": "@grouparoo/spec-helper",
  "description": "Shared Utilities for testing Grouparoo and Grouparoo Plugins",
  "version": "0.2.6-alpha.0",
  "license": "MPL-2.0",
  "private": false,
  "publishConfig": {
    "access": "public"
  },
  "engines": {
    "node": ">=12.0.0 <16.0.0"
  },
  "main": "dist/index",
  "homepage": "https://www.grouparoo.com",
  "bugs": {
    "url": "https://github.com/grouparoo/grouparoo/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/grouparoo/grouparoo.git",
    "directory": "plugins/@grouparoo/spec-helper"
  },
  "scripts": {
    "prepare": "rm -rf dist && tsc --declaration",
    "test": "jest",
    "pretest": "npm run lint && npm run prepare",
    "lint": "prettier --check src ",
    "watch": "tsc --watch"
  },
  "dependencies": {
    "axios": "0.21.1",
    "faker": "5.4.0",
    "nock": "13.0.10",
    "sequelize": "6.5.0",
    "uuid": "8.3.2"
  },
  "devDependencies": {
    "@grouparoo/core": "0.2.6-alpha.0",
    "@types/faker": "*",
    "@types/jest": "*",
    "@types/node": "*",
    "actionhero": "25.0.7",
    "jest": "26.6.3",
    "prettier": "2.2.1",
    "ts-jest": "26.5.3",
    "typescript": "4.2.2"
  },
  "gitHead": "d87e6adcefcc3e55d671121157b8eda1ae89f22a"
}
