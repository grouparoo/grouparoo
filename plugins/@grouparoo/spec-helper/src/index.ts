export * from "./lib/specHelper";
export * from "./lib/workflows/import";
export * from "./lib/relaxedSnapshot";
export * from "./lib/integrationSpecHelper";
export * from "./lib/cliSpecHelper";
