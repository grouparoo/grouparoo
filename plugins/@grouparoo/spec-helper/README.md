# @grouparoo/spec-helper

Shared Utilities for testing [Grouparoo](https://www.grouparoo.com) and Grouparoo Plugins. Learn more at https://www.grouparoo.com/docs/running/testing.

A typical test file's shape:

```ts
import { helper } from "@grouparoo/spec-helper";

describe("test suite", () => {
  helper.grouparooTestServer({ truncate: true, enableTestPlugin: true });

  test("it works", () => {
    expect(1 + 1).toBe(2);
  });
});
```

Since you are asking Jest/TS-Jest to compile a Typescript file from within `node_modules`, you will need to add the following to your jest config (usually `jest.config.js`)

```js
transformIgnorePatterns: ["node_modules/(?!(@grouparoo)/)"];
```

## Notes

1. You _must_ require `@grouparoo/spec-helper` as the first import in your test files.
2. If you want to inject a plugin to the running Grouparoo test instance, use the `GROUPAROO_INJECTED_PLUGINS` environment variable:

```ts
import path from "path";
process.env.GROUPAROO_INJECTED_PLUGINS = JSON.stringify({
  "@grouparoo/bigquery": { path: path.join(__dirname, "..", "..") },
});
```
