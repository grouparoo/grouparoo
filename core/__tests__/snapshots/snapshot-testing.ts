import { helper, relaxedSnapshot } from "@grouparoo/spec-helper";
import {
  GrouparooRecord,
  Group,
  Destination,
  GrouparooPlugin,
  Source,
} from "../../src";
import { api } from "actionhero";

describe("test grouparoo records", () => {
  helper.grouparooTestServer({ truncate: true, enableTestPlugin: true });

  describe("within a mock environment", () => {
    let record: GrouparooRecord;
    let group: Group;
    let destination: Destination;

    beforeAll(async () => {
      // we need to ensure that we have static results for the import to snapshot
      const testPlugin: GrouparooPlugin = api.plugins.plugins.find(
        (a) => a.name === "@grouparoo/test-plugin"
      );
      const testPluginConnection = testPlugin.connections.find(
        (c) => c.name === "test-plugin-import"
      );
      testPluginConnection.methods.recordProperty = async ({ property }) => {
        const data = {
          userId: [100],
          isVIP: [true],
          email: [`example@example.com`],
          firstName: ["Mario"],
          lastName: ["Mario"],
          ltv: [100.0],
          lastLoginAt: [new Date(0)],
          purchases: ["hat", "mushroom"],
          purchaseAmounts: [100, 200],
        };

        return data[property.key];
      };
    });

    beforeAll(async () => {
      // make properties
      await helper.factories.properties();
      const source = await Source.findOne();

      // make a group
      group = await helper.factories.group();
      await group.update({ name: "test group" });
      await group.setRules([
        { key: "email", match: "%@example.com", operation: { op: "like" } },
      ]);

      // make a destination
      destination = await helper.factories.destination();
      await destination.update({ name: "test destination" });
      await destination.setOptions({ table: "users" });
      await destination.setMapping({
        "primary-id": "userId",
        email: "email",
        ltv: "ltv",
        isVIP: "isVIP",
      });
      await destination.setDestinationGroupMemberships({
        [group.id]: "remote_group_name",
      });
      await destination.updateTracking("group", group.id);

      // make ready
      await group.update({ state: "ready" });
      await destination.update({ state: "ready" });

      // make the record
      const response =
        await GrouparooRecord.findOrCreateByUniqueRecordProperties(
          { userId: ["person1@example.com"] },
          source
        );
      record = response.record;

      // import & export the record
      await record.sync();
    });

    test("record properties match the snapshot", async () => {
      const properties = await record.simplifiedProperties();
      expect(properties).toMatchSnapshot();
    });

    test("groups match the snapshot", async () => {
      const groups = await record.$get("groups");
      const groupApiData = await Promise.all(groups.map((g) => g.apiData()));

      expect(groupApiData.length).toEqual(1);

      expect(groupApiData[0]).toMatchSnapshot({
        id: expect.stringMatching(/^grp_/),
        createdAt: expect.any(Number),
        updatedAt: expect.any(Number),
        nextCalculatedAt: null,
      });
    });

    test("potential exports match the snapshot", async () => {
      const _exports = await record.export(true, [], false);
      const exportApiData = await Promise.all(
        _exports.map((e) => e.apiData(false))
      );
      expect(exportApiData.length).toEqual(1);

      expect(exportApiData[0]).toMatchSnapshot({
        recordId: expect.stringMatching(/^rec_/),
        state: "pending",
        startedAt: expect.any(Number),
        sendAt: expect.any(Number),
        newRecordProperties: expect.objectContaining({
          email: expect.stringMatching(/@example.com/),
          "primary-id": expect.any(Number),
          isVIP: true,
          ltv: 100,
        }),
        oldRecordProperties: expect.objectContaining({
          email: expect.stringMatching(/@example.com/),
          "primary-id": expect.any(Number),
          isVIP: true,
          ltv: 100,
        }),
      });
    });

    test("the record can be snapshot-ed", async () => {
      const snapshot = await record.snapshot();
      expect(snapshot).toMatchSnapshot(relaxedSnapshot(snapshot));
    });
  });
});
