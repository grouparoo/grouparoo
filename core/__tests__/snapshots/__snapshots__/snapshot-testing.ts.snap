// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test grouparoo profiles within a mock environment groups match the snapshot 1`] = `
Object {
  "calculatedAt": null,
  "createdAt": Any<Number>,
  "id": StringMatching /\\^grp_/,
  "locked": null,
  "matchType": "all",
  "name": "test group",
  "nextCalculatedAt": Any<Number>,
  "profilesCount": 1,
  "rules": Array [
    Object {
      "key": "email",
      "match": "%@example.com",
      "operation": Object {
        "description": "is like (case sensitive)",
        "op": "like",
      },
      "relativeMatchDirection": null,
      "relativeMatchNumber": null,
      "relativeMatchUnit": null,
      "topLevel": false,
      "type": "email",
    },
  ],
  "state": "ready",
  "type": "calculated",
  "updatedAt": Any<Number>,
}
`;

exports[`test grouparoo profiles within a mock environment potential exports match the snapshot 1`] = `
Object {
  "completedAt": null,
  "createdAt": null,
  "destination": null,
  "destinationName": "test destination",
  "errorLevel": undefined,
  "errorMessage": undefined,
  "force": true,
  "hasChanges": true,
  "id": null,
  "newGroups": Array [
    "remote_group_name",
  ],
  "newProfileProperties": ObjectContaining {
    "email": StringMatching /@example\\.com/,
    "isVIP": true,
    "ltv": 100,
    "primary-id": Any<Number>,
  },
  "oldGroups": Array [
    "remote_group_name",
  ],
  "oldProfileProperties": ObjectContaining {
    "email": StringMatching /@example\\.com/,
    "isVIP": true,
    "ltv": 100,
    "primary-id": Any<Number>,
  },
  "profileId": StringMatching /\\^pro_/,
  "retryCount": 0,
  "sendAt": Any<Number>,
  "startedAt": Any<Number>,
  "state": "pending",
  "toDelete": false,
}
`;

exports[`test grouparoo profiles within a mock environment profile properties match the snapshot 1`] = `
Object {
  "email": Array [
    "example@example.com",
  ],
  "firstName": Array [
    "Mario",
  ],
  "isVIP": Array [
    true,
  ],
  "lastLoginAt": Array [
    1970-01-01T00:00:00.000Z,
  ],
  "lastName": Array [
    "Mario",
  ],
  "ltv": Array [
    100,
  ],
  "purchaseAmounts": Array [
    100,
    200,
  ],
  "purchases": Array [
    "hat",
    "mushroom",
  ],
  "userId": Array [
    100,
  ],
}
`;

exports[`test grouparoo profiles within a mock environment the profile can be snapshot-ed 1`] = `
Object {
  "exports": Array [
    Object {
      "completedAt": null,
      "createdAt": null,
      "destination": null,
      "destinationName": "test destination",
      "errorLevel": undefined,
      "errorMessage": undefined,
      "force": true,
      "hasChanges": true,
      "id": null,
      "newGroups": Array [
        "remote_group_name",
      ],
      "newProfileProperties": Object {
        "email": "example@example.com",
        "isVIP": true,
        "ltv": 100,
        "primary-id": 100,
      },
      "oldGroups": Array [
        "remote_group_name",
      ],
      "oldProfileProperties": Object {
        "email": "example@example.com",
        "isVIP": true,
        "ltv": 100,
        "primary-id": 100,
      },
      "profileId": Any<String>,
      "retryCount": 0,
      "sendAt": Any<Number>,
      "startedAt": Any<Number>,
      "state": "pending",
      "toDelete": false,
    },
  ],
  "groups": Array [
    Object {
      "calculatedAt": null,
      "createdAt": Any<Number>,
      "id": Any<String>,
      "locked": null,
      "matchType": "all",
      "name": "test group",
      "nextCalculatedAt": Any<Number>,
      "profilesCount": 1,
      "rules": Array [
        Object {
          "key": "email",
          "match": "%@example.com",
          "operation": Object {
            "description": "is like (case sensitive)",
            "op": "like",
          },
          "relativeMatchDirection": null,
          "relativeMatchNumber": null,
          "relativeMatchUnit": null,
          "topLevel": false,
          "type": "email",
        },
      ],
      "state": "ready",
      "type": "calculated",
      "updatedAt": Any<Number>,
    },
  ],
  "properties": Object {
    "email": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "email",
      "unique": true,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        "example@example.com",
      ],
    },
    "firstName": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        "Mario",
      ],
    },
    "isVIP": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "boolean",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        true,
      ],
    },
    "lastLoginAt": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "date",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        1970-01-01T00:00:00.000Z,
      ],
    },
    "lastName": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        "Mario",
      ],
    },
    "ltv": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "float",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        100,
      ],
    },
    "purchaseAmounts": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": true,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "float",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        100,
        200,
      ],
    },
    "purchases": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": false,
      "isArray": true,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "string",
      "unique": false,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        "hat",
        "mushroom",
      ],
    },
    "userId": Object {
      "confirmedAt": Any<Date>,
      "createdAt": Any<Date>,
      "id": Any<String>,
      "identifying": true,
      "isArray": false,
      "startedAt": null,
      "state": "ready",
      "stateChangedAt": Any<Date>,
      "type": "integer",
      "unique": true,
      "updatedAt": Any<Date>,
      "valueChangedAt": Any<Date>,
      "values": Array [
        100,
      ],
    },
  },
}
`;
